midterm_project/project_mihaela_bianca/shopping_list.py

```diff
@@ -0,0 +1,286 @@
+import random
+import pprint
+
+from copy import deepcopy
+
+archived_list = []
+
+pp = pprint.PrettyPrinter(indent=4)
+
+__header = r'''
+   ______________________________
+ / \                             \.
+|   |                            |.
+ \_ |                            |.
+    |                            |.
+'''
+__footer = r'''
+    |                            |.
+    |   _________________________|___
+    |  /                            /.
+    \_/____________________________/.
+'''
+
+
+class PrettyPrinterMixin:
+
+    def __str__(self):
+        dash = '*' * 30
+
+        if hasattr(self, 'name'):
+            header_name = '{} for {}'.format(self.__class__.__name__, self.name)
+        else:
+            header_name = self.__class__.__name__
+
+        header = '{} contains:'.format(header_name)
+
+        result = '\n\n{} \n{}\n{} \n '.format(dash, header, dash)
+        for index, ingred in enumerate(self.ingredients, start=1):
+            result += '\n{}. {} : {} \n '.format(index, ingred.ljust(22, ' '), self.ingredients[ingred])
+        result += '\n{}'.format(dash)
+        return result
+
+
+class Recipe(PrettyPrinterMixin):
+
+    def __init__(self, name, ingredients):
+        """
+        Ingredients should be a dictionary
+        {egg:3, milk:3, flour:2, sugar:1}
+        """
+        self.name = name
+        self.ingredients = ingredients
+
+        if len(self.ingredients) < 4:
+            raise Exception('The ingredients number should be  at least 4 !')
+
+    def __iter__(self):
+        yield from self.ingredients.items()
```

Comment:
nice use of yield. An alternative here could be `iter(self.ingredients.items())`

Reply:
modified